pdns/cachecleaner.hh

```diff
@@ -167,43 +176,51 @@ uint64_t pruneMutexCollectionsVector(C& container, std::vector<T>& maps, uint64_
         ++i;
       }
 
-      if (toTrim && erased >= toTrim / maps_size)
+      if (erased >= toTrimForThisShard) {
         break;
+      }
 
-      if (lookedAt > lookAt / maps_size)
+      if (lookedAt >= toScanForThisShard) {
         break;
+      }
     }
     totErased += erased;
-    if (toTrim && totErased >= toTrim)
-      break;
   }
 
   if (totErased >= toTrim) { // done
     return totErased;
   }
 
   toTrim -= totErased;
+  cacheSize -= totErased;
 
-  while (true) {
-    size_t pershard = toTrim / maps_size + 1;
-    for (auto& content : maps) {
-      auto mc = content.lock();
-      mc->invalidate();
-      auto& sidx = boost::multi_index::get<S>(mc->d_map);
-      size_t removed = 0;
-      for (auto i = sidx.begin(); i != sidx.end() && removed < pershard; removed++) {
-        container.preRemoval(*mc, *i);
-        i = sidx.erase(i);
-        --content.d_entriesCount;
-        totErased++;
-        toTrim--;
-        if (toTrim == 0) {
-          return totErased;
-        }
+  /* it was not enough, so we need to remove entries that are not expired,
+     still using the LRU */
+
+  double remainingShards = numberOfShards;
+  const auto factor = static_cast<double>(numberOfShards) / static_cast<double>(cacheSize);
+  for (auto& content : maps) {
+    auto shard = content.lock();
+    const auto shardSize = shard->d_map.size();
+    const uint64_t toTrimForThisShard = std::round(factor * toTrim * shardSize / remainingShards);
```

Comment:
I'm not sure the computation is still correct for the last shards, as the number of entries (`cacheSize`) from which the ratio is computed is not updated. I wonder if taking the value `toTrim` before that loop times `cacheSize` before the loop divided by the shard size would not be more accurate?

Reply:
Hmm I think I messed up when cleaning the code before commit. Will revisit.

Reply:
Dividing by shard size is not correct in any case, as we want ro remove more from a shard if it is big

Reply:
So the issue with `const uint64_t toTrimForThisShard = std::ceil(1.0 * toTrim * shardSize / cacheSize);`
is that it overshoots (cleans too much) because of the `ceil`. A `round` undershoots in some cases. I was trying to compensate for that by updating `toTrim` and `cacheSize` in the loop but obviously messed up.

I need to think about if overshooting (cleaning too much) is bad (it is limited by the number of shards anyway).